_posts/2020-07-20-github-review-and-merge.md

```diff
@@ -0,0 +1,131 @@
+by the tooling of a project to ease and help its implementation.
+
+This post is looking to discuss the last two steps:
+- Reviews are essential to a healthy codebase, providing correctness, quality
+  and a learning opportunity with focus on team collaboration.
+- Merging combines work into the collaborative project, providing a sync point
+  and keeping the work history.
+  Codebase change history is crucial for maintaining and understanding the
+  code. It provides context to changes with proper reasoning and relevant data.
+
+# Reviewable Change
+Write code with your reviewers in mind.
+
+- Small is greater than Big:
+  Small changes are easy to review and simple to explain.
+  - Keep PR/s small:
+    Discussions are performed on PR level in GitHub, therefore, less commits
+    in a PR encourages focused discussions.
+    This obviously comes with a cost: Increased test runs and stacking PR/s
+    one over the other in a dependency chain.
+  - Keep commits small:
+    Small commit messages are focused on a specific intent (described in the
+    commit message), giving the reviewer a smaller domain to focus on. It also
+    provides an insight of the change direction, telling a story.
+- Commit messages provide intent and context to changes for the reviewer.
+- Changes performed due to review requests, should be easily tracked by the
+  reviewer (this is one of the main pain points with GitHub review flow).
+  This issue is elaborated in the
+  [Track Requested Changes](#track-requested-changes)
+
+# Merge Methods
+The method chosen to merge changes into the collaborative branch (e.g. master)
+impacts mainly the change history.
+
+GitHub (through git) provides 3 main
+[methods](https://docs.github.com/en/github/administering-a-repository/about-merge-methods-on-github)
+of merging a PR:
+- Merge
+- Squash
+- Rebase (my favorite)
+
+It is worth mentioning that in GitHub, a PR represents a collection of changes
+that are merged together. Individual commits cannot be merged (unless they
+are wrapped by a PR, in a 1:1 relation).
+
+The merge and rebase methods are preserving the PR commit history intact.
+The squash method on the other hand is combining all PR commits into a single
+commit (recording all commit messages into a single commit message)
+automatically.
+
+The merge method preserves the feature branch, allowing future usage of the
+changes included in it (e.g. backporting or sharing with multiple branches).
+On the other hand, the rebase method places the changes one by one on the
+target branch, keeping a cleaner linear log history but without the changes
+original grouping.
+
+If you value your change history and commit message, do not use squash.
+
+If you intend to merge changes into multiple branches frequently or your
+individual commits are not self contained (i.e. they cannot stand alone without
+the other changes in the PR when tested), use the merge method.
```

Comment:
This is worth describing in more detail since you are talking about GitHub.

GitHub frames the whole PR as one huge diff, so it is not guaranteed, reviewers will consider each commit separately. And the CI runs on the tip of the branch. So with `rebase`, you are risking a lot, `merge` is the way to go with shitty GitHub

Reply:
Done